### ISSUE TITLE
Implicit read-only-ness doesn't seem to have been implemented for object-constructor-expressions 

### ISSUE THREAD

**Issue by MaryamZi** [ballerina-platform/ballerina-lang#26935]
**Description:**
The [spec](https://ballerina.io/spec/lang/master/#section_6.8.1) says

> The object-constructor-expr is implicitly read-only if the applicable contextually expected type is a subtype of readonly or if the type-reference is present and references a type that is a subtype of readonly. If object-constructor-expr is implicitly read-only, then the object will be constructed with its read-only bit on.

**Steps to reproduce:**
But the following assignments aren't allowed.

```ballerina
readonly class Foo {
    int i;

    function init(int i) {
        self.i = i;
    }
}

readonly class Bar {
    int i = 1;
    string s = "hello";
}

public function main() {
    Foo f1 = object { // error: incompatible types: expected 'Foo', found 'object { int i; string s; }'
        int i = 1;
        string s = "hello";
    }; 

    Foo f2 = object Bar { // error: incompatible types: expected 'Foo', found 'object { int i; string s; }'
        int i = 1;
        string s = "hello";
    }; 

    // Works
    // Foo f3 = new Bar();
}
```

Also the following print `false`:

```ballerina
import ballerina/io;

readonly class Foo {
    int i;

    function init(int i) {
        self.i = i;
    }
}

public function main() {
    var x = object Foo {
        int i = 1;
        string s = "hello";
    }; 

    any a = x;
    io:println(a is readonly); // false
    io:println(a is Foo); // false
}
```

**Affected Versions:**
slp5

**Related Issues (optional):**
Blocker for https://github.com/ballerina-platform/ballerina-lang/issues/26704

**Comment by MaryamZi** [comment]
This requires changes to the `ImmutableTypeCloner` to define a new class for an object instead of an object type definition, for the class that represents the object constructor expression.

**Comment by MaryamZi** [comment]
Depends on https://github.com/ballerina-platform/ballerina-lang/issues/26969

**Comment by MaryamZi** [comment]
The implicit readonly-ness of an object constructor with a type-reference will be fixed when fixing #26969.

This issue will focus on the implicit readonly-ness of an object constructor where the contextually expected type is a subtype of `readonly`.

**Comment by MaryamZi** [comment]
> This requires changes to the `ImmutableTypeCloner` to define a new class for an object instead of an object type definition, for the class that represents the object constructor expression.

Since a class definition is created per object constructor expression, we can mark the original class as `readonly` instead of creating a new class.
